[WIP] Added functions setenv(), unsetenv(), & putenv() and the corresponding test suite

LeeTibbert opened the pull request:
This was work done in preparation for fixing issue #1167, TimeSuite failure.
@eatkins has, thankfully, created a PR for that issue. This work may
have value in other contexts.

The new StdlibSuite.scala tests only the functions introduced in this commit.
It does not test all of stdlib.scala, but it is a beginning of a more
comprehensive suite.

getenv() & friends are defined as being thread-unsafe.  The test suite
must be run in a single threaded environment. Only the testing
framework, not the test suite, can guarantee that.  

I will create an issue for the testing framework. More details there.

Tested on both 64-bit & X86 32-bit machines.

ekrich commented:
`setenv` and `unsetenv` are already in master. I did not add `putenv` as it didn't seem to be favored.
https://github.com/scala-native/scala-native/pull/1151

LeeTibbert commented:
@ekrich, @eatkins,

Sorry if I added confusion.  When I started out trying to fix #1167, I found getenv() in
scalanative/native/stdlib.scala, so I decided that setenv & friends should go there.
The former seems to have a more complete list of functions from stdlib.h but you
must have had a reason for creating the latter.

Given your information above, l just found scalanative/posix/stdlib.scala in my
64 bit marshaling repository. As you say, which does, indeed, contain setenv & friends.

I did not find a test suite for either stdlib.scala.  I am currently chasing what appears
to be a memory corruption bug on X86_64 with mktime(). I do not yet have enough
evidence to indict but it is sucking up even more of my time. Nasty if true.  If you like,
after I settle that issue to my satisfaction, I could rework the StdlibSuite.scala to
apply to posix/stdlib.scala.  Probably cleaner if I close this PR and open another one
when ready.

Your thoughts?

I am no fan of putenv() and writing the tests for it modifying the buffer
was an exercise in attempting to have compassion for people who
have other views of it.  I added it only for completeness. From a search
of the InterWeb, the discussion about putenv() is long, partisan, & painful.
It appears that there is, still, a group of people who like the memory
control that an explicit & free of the tm buffer allows.  They are concerned
the with, apparently strong or at least unknown, possibility of setenv() having
a memory leak.  It seemed cheaper all around to add it for completeness now
rather than have to respond to an issue and wait a release cycle.  The people
who like it really seem to like it and, to my knowledge, has not been deprecated.
The test suite quotes Murray Gell-Mann about things not being prohibited
must occurring.

Your call.  Please advise.  Thank you.

Lee

ekrich commented:
@LeeTibbert I had a PR,  https://github.com/scala-native/scala-native/pull/1059 now closed, trying to do too many things some of which reviewed poorly. I looked closer and decided to do the first part, the POSIX `stdlib` bindings and then attack other portions. What I noticed is that `setenv` and friends where in the POSIX side versus the `C stdlib`. I also noticed that for just basic `@extern` functions there were no tests. I guess the exception is like the `complex` lib where we have C backing code and the `extern`ed methods are not just a direct translation and more work needs to be done. 

I think the `time` regression can be fixed now but feel free to add `putenv` if you wish. My approach is just to add what is needed partly out of laziness but partly because to support Windows is going to be hard so less is better. Anyway, that is what I was thinking.

LeeTibbert commented:
This WIP PR has served its review purpose. Close to reduce cognitive load, all around.

I learned setenv() and unsetenv() already exist and putenv() alone does not have a favorable
cost/benefit ratio, given its baggage: loved by those who love it (not me) and disparaged by
those who do not.
